internal/monitors/expvar/expvar.go

```diff
@@ -164,90 +164,113 @@ func (m *Monitor) findKeys(pattern string, aMap map[string]interface{}) []string
 	return keys
 }
 
-// setDps sets metric values and dimensions by traversing jsonValue recursively using keys derived by regex matching
-// configured JSON paths. When setDps encounters and array type JSON value it clones the datapoint argument and adds
+// addDps sets metric values and dimensions by traversing jsonValue recursively using keys derived by regex matching
+// configured JSON paths. When addDps encounters and array type JSON value it clones the datapoint argument and adds
 // dimension to hold the array index.
-func (m *Monitor) setDps(dps map[string][]*datapoint.Datapoint, dp *datapoint.Datapoint, jsonKey string, jsonKeyPatternIndex int, jsonValue interface{}, metricConfig *MetricConfig) {
-	dp.Metric = joinWords(snakeCaseSlice([]string{dp.Metric, jsonKey}), ".")
-	jsonKeyPatterns, nextJSONKeyPatternIndex := m.metricsKeys[metricConfig.Name], jsonKeyPatternIndex+1
-	if jsonKeyPatternIndex >= len(jsonKeyPatterns) {
-		m.logger.Errorf("failed to find metric value in path: %s", metricConfig.JSONPath)
-		return
-	}
-	switch jsonValue := jsonValue.(type) {
+func (m *Monitor) addDps(dps map[string][]*datapoint.Datapoint, dp *datapoint.Datapoint, keys []string, keyFromPathIndex int, value interface{}, conf *MetricConfig) {
+	keysFromPath, pathKeysLen, nextKeyFromPathIndex := m.keysFromPaths[conf.Name], len(m.keysFromPaths[conf.Name]), keyFromPathIndex+1
+	atPathEnd := nextKeyFromPathIndex == pathKeysLen
+	switch value := value.(type) {
 	case map[string]interface{}:
-		nextJSONKeyPattern := jsonKeyPatterns[nextJSONKeyPatternIndex]
-		dpCopy, metric, dims := dp, dp.Metric, dp.Dimensions
-		for i, nextJSONKey := range m.findKeys(nextJSONKeyPattern, jsonValue) {
+		if atPathEnd {
```

Comment:
This function is getting pretty unwieldy.  Also it's pretty confusing having a function called `addDps` and another called `addDp` that do radically different things.